includes/sanitizers/class-amp-meta-sanitizer.php

```diff
@@ -0,0 +1,333 @@
+			} elseif ( 'viewport' === $element->getAttribute( 'name' ) ) {
+				$this->meta_tags[ self::TAG_VIEWPORT ][] = $element;
+			} elseif ( 'amp-script-src' === $element->getAttribute( 'name' ) ) {
+				$this->meta_tags[ self::TAG_AMP_SCRIPT_SRC ][] = $element;
+			} else {
+				$this->meta_tags[ self::TAG_OTHER ][] = $element;
+			}
+		}
+
+		$this->ensure_charset_is_present( $charset );
+
+		if ( ! $this->is_correct_charset() ) { // phpcs:ignore Generic.CodeAnalysis.EmptyStatement
+			// @TODO Re-encode the content into UTF-8.
+			// ... sure?
+		}
+
+		$this->ensure_viewport_is_present();
+
+		$this->process_amp_script_meta_tags();
+
+		$this->re_add_meta_tags_in_optimized_order();
+	}
+
+	/**
+	 * Ensure that the <head> element is present in the document.
+	 *
+	 * @return DOMElement The document's <head> element.
+	 */
+	protected function ensure_head_is_present() {
+		$head = $this->dom->getElementsByTagName( 'head' )->item( 0 );
+
+		if ( ! $head ) {
+			$head = $this->dom->createElement( 'head' );
+			$head = $this->dom->documentElement->insertBefore( $head, $this->dom->documentElement->firstChild );
+		}
+
+		return $head;
+	}
+
+	/**
+	 * Detect the charset of the document.
+	 *
+	 * @return string|false Detected charset of the document, or false if none.
+	 */
+	protected function detect_charset() {
+		$charset = false;
+
+		// Check for HTML 4 http-equiv meta tags.
+		$http_equiv_tag = $this->xpath->query( '//meta[ @http-equiv and @content ]' )->item( 0 );
+		if ( $http_equiv_tag ) {
+			$http_equiv_tag->parentNode->removeChild( $http_equiv_tag );
+
+			// Check for the existence of a proper charset attribute first.
+			$charset = $http_equiv_tag->getAttribute( 'charset' );
+			if ( ! $charset ) {
+				// If not, check whether the charset is included with the content type, and use that.
+				$content = $http_equiv_tag->getAttribute( 'content' );
+
+				$matches = [];
+				if ( preg_match( '/;\s*charset=(?<charset>[^;]+)/', $content, $matches ) && ! empty( $matches['charset'] ) ) {
+					$charset = $matches['charset'];
+				}
+			}
+		}
+
+		// Check for HTML 5 charset meta tag. This overrides the HTML 4 charset.
+		$charset_tag = $this->xpath->query( '//meta[ @charset ]' )->item( 0 );
+		if ( $charset_tag ) {
+			$charset_tag = $charset_tag->parentNode->removeChild( $charset_tag );
+			$charset     = $charset_tag->getAttribute( 'charset' );
+		}
+
+		return $charset;
+	}
+
+	/**
+	 * Always ensure that we have an HTML 5 charset meta tag.
+	 *
+	 * The charset defaults to utf-8, which is also what AMP requires.
+	 *
+	 * @param string|false $charset Optional. Charset that was already detected. False if none. Defaults to false.
+	 */
+	protected function ensure_charset_is_present( $charset = false ) {
+		if ( ! empty( $this->meta_tags[ self::TAG_CHARSET ] ) ) {
+			return;
+		}
+
+		$this->meta_tags[ self::TAG_CHARSET ][] = $this->create_charset_element( $charset ?: static::AMP_CHARSET );
```

Comment:
I think this can just always use `static::AMP_CHARSET` because this is the only thing that AMP allows (`utf-8`).

Reply:
Yes, but only if we do indeed transform the document first. Otherwise, we're not only in the wrong charset, but we've also lost the information about what the actual charset is => broken^2. :)